spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,323 @@
+    StreamingOffset startingOffset;
+    if (startSnapshotId != null) {
+      startingOffset = new StreamingOffset(startSnapshotId, 0, true, false);
+    } else {
+      List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+      if (snapshotIds.isEmpty()) {
+        // there's no snapshot currently.
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(snapshotIds.get(snapshotIds.size() - 1), 0, true, false);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    // Spark will invoke setOffsetRange more than once. If this is already calulated, use the cached one to avoid
+    // calculating again.
+    if (cachedPendingBatches == null || !cachedPendingBatches.first().equals(start)) {
+      this.cachedPendingBatches = Pair.of(start, getChangesWithRateLimit(start, maxSizePerBatch));
+    }
+
+    List<MicroBatch> batches = cachedPendingBatches.second();
+    MicroBatch lastBatch = batches.isEmpty() ? null : batches.get(batches.size() - 1);
+
+    if (lastBatch == null) {
+      return start;
+    } else {
+      boolean isStarting = lastBatch.snapshotId() == start.snapshotId() && start.shouldScanAllFiles();
+      return new StreamingOffset(lastBatch.snapshotId(), lastBatch.endFileIndex(), isStarting,
+          lastBatch.lastIndexOfSnapshot());
+    }
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("checkstyle:HiddenField")
+  List<MicroBatch> getChangesWithRateLimit(StreamingOffset startOffset, long maxSize) {
+    List<MicroBatch> batches = Lists.newArrayList();
+    long currentLeftSize = maxSize;
+    MicroBatch lastBatch = null;
+
+    checkOverwrite(table.snapshot(startOffset.snapshotId()));
+    if (shouldGenerateFromStartOffset(startOffset)) {
+      currentLeftSize -= generateMicroBatch(startOffset.snapshotId(), startOffset.index(),
+          startOffset.shouldScanAllFiles(), currentLeftSize, batches);
+      lastBatch = batches.get(batches.size() - 1);
+    }
+
+    // Current snapshot can already satisfy the size needs.
+    if (currentLeftSize <= 0L || (lastBatch != null && !lastBatch.lastIndexOfSnapshot())) {
+      return batches;
+    }
+
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+    if (currentSnapshotId == startOffset.snapshotId()) {
+      // the snapshot of current offset is already the latest snapshot of this table.
+      return batches;
+    }
+
+    ImmutableList<Long> snapshotIds = ImmutableList.<Long>builder()
+        .addAll(SnapshotUtil.snapshotIdsBetween(table, startOffset.snapshotId(), currentSnapshotId))
+        .build()
+        .reverse();
+
+    for (Long id : snapshotIds) {
+      Snapshot snapshot = table.snapshot(id);
+      checkOverwrite(snapshot);
+      if (!isAppend(snapshot)) {
```

Comment:
Probably worth to mention in PR or doc later on limitations due to the incremental read. 

I see two cases: 1. Cannot read from snapshot with overwrite operation. 2. Deleted files from table may be still read afterwards.

Is my understanding correct?

Reply:
Yes, correct, it is the same behavior to incremental scan.